### PR TITLE
feat: Add basic typing support

### PR DESCRIPTION
**Pull request by last-partizan**
This PR adds basic type hinting support

Refs #468

### CODE REVIEW

**Review by jeffwidman — APPROVED** [review]
Thanks!

**Review by rbarrois — COMMENTED** [review]
That's interesting, thanks!

I've got a few questions though:
- How could we ensure that the annotations are actually working and useful? Can we test those?
- There doesn't seem to be a simple way to link the `Generic[T]` with the `class Meta: model`

**Comment by last-partizan** [comment]
@rbarrois

hmm, i think, to properly test type annotations, we should add pyright and/or mypy to CI, run it over example file, and compare that to expected output.

But there is simpler option, we can do something like this:

```
def x() -> int:
    ...

assert x.__annotations__ == {"returns": int}
```

In theory, this should be enough to ensure we get expected annotations.

as for linking with `Meta: model`, yes, i don't see any way to do it.

Also, i see some failing CI checks, i'll try to fix it.

**Comment by last-partizan** [comment]
Oh, looks like all errors is from 3.6 version, lefetime of which is ended three weeks ago.

Maybe better solution in this case, is to remove old version from CI? But i'll try to find other solutions, maybe any ideas?

**Comment by last-partizan** [comment]
I haven't found easy workarounds for this, so i created PR for removing 3.6 from CI, and adding 3.10 and django-4.0.

**Comment by mathieutu** [comment]
Thanks, an other huge step would be to type the kwargs. 
To be able to do `UserFactory.create(xxx=yy)` and have autocompletion on `xxx`. 

I have no idea on how to do that, and even if it could be possible with python? 🤔

**Comment by last-partizan** [comment]
I think, it could be implemented. Python 3.10 added `ParamSpec`, and it could be used with older versions of python via typing-extensions, so i think it could be possible.

**Comment by last-partizan** [comment]
I found a way to properly type even calls to factory, turned out it was pretty simple.

Now types are correct with both possible ways to call a factory:

```
BookFactory() -> Book
BookFactory.create() -> Book
```

But, this all requires python 3.6 at least.

**Comment by mathieutu** [comment]
> I think, it could be implemented. Python 3.10 added `ParamSpec`, and it could be used with older versions of python via typing-extensions, so i think it could be possible.

I would love to see that!

**Comment by last-partizan** [comment]
> I found a way to properly type even calls to factory, turned out it was pretty simple.

Turned out it doesn't work. Reverted changes back to only typing `create`, `build`, etc...

Now i'm trying to test it.

**Comment by last-partizan** [comment]
As for testing type annotations, i found one way to do it:

And it looks like this: https://github.com/smarie/python-decopatch/pull/27/files

- Install `pyright`
- Run it in json mode, compare results to snapshot

I could add similar tests for this package. (In this PR or next, preferrably next)

**Comment by last-partizan** [comment]
I fund myself today thinking it would be great to have types in my factories. Can we merge this, or we need to add tests first?

**Comment by youtux** [comment]
I started a [types-factory-boy](https://github.com/youtux/types-factory-boy) to add type annotations to factory boy. It is in early stage, and it most probably needs a plugin to automatically discover and create type annotations automatically (otherwise you would have to type annotate basically every attribute), but I think it's a start.

**Comment by flaeppe** [comment]
When it comes to testing types and type annotations. I'd suggest: https://github.com/TypedDjango/pytest-mypy-plugins

Although it's a `pytest` plugin, so those tests needs to be run with `pytest`. Should be possible to add that as an individual job in CI etc.

Have a look at `django-stubs` https://github.com/typeddjango/django-stubs/tree/master/tests/typecheck for reference of how to use it and/or set it up.

**Comment by last-partizan** [comment]
@youtux great project, thank you.

After pyright fixed a bug with invalid __init__ annotations and default values in 1.1.264, default factory_boy is throwing type errors sometimes (Maybe declaration for example). 

Your project fixed it.

**Comment by jeffwidman** [comment]
Is this ready for review again?

**Comment by last-partizan** [comment]
In basic form, yes.

But, we have https://github.com/youtux/types-factory-boy now, and i can try to merge it into this project with [merge_pyi](https://github.com/google/pytype/tree/main/pytype/tools/merge_pyi).

But, i think it's better to leave this PR in minimal form, maybe add tests for types in the next PR. And then - merge types-factory-boy for more complete type coverage.

**Comment by youtux** [comment]
I think there was some misunderstanding here, as I don't understand @last-partizan message about merging [types-factory-boy](https://github.com/youtux/types-factory-boy) into this codebase.

`types-factory-boy` contains annotations for factory_boy lib (which I don't control), not for pytest_factoryboy (that I manage).

**Comment by youtux** [comment]
I made some progress in types-factory-boy. [Now](https://github.com/youtux/types-factory-boy/commit/35ef612ada5ee4fc29cf483e268e695968e57e52) there is a mypy plugin that automatically adds the typevar information of the `Factory[T]` baseclass.

So, instead of writing this:
```python
class AuthorFactory(Factory[Author]):
    class Meta:
        model = Author
reveal_type(AuthorFactory.create())  # foo.Author
```

we can write this:

```python
class AuthorFactory(Factory):  # No need to specify the typevar for Factory[...], it's inferred from Meta.model
    class Meta:
        model = Author
reveal_type(AuthorFactory.create())  # foo.Author
```

**Comment by palgorhythm** [comment]
Any updates on whether this PR will be merged?

**Comment by last-partizan** [comment]
Updated it to include tests for typing.

**Comment by last-partizan** [comment]
But i need some help for running type check tests only on python, not pypy.

Here's my attempt: https://github.com/last-partizan/factory_boy/pull/2

But it seems doesn't run any of py..-typecheck envs at all.

https://github.com/last-partizan/factory_boy/actions/runs/4702801405/jobs/8340503694

Any ideas why?

**Comment by last-partizan** [comment]
@rbarrois please, take a look. I've added tests for types based on [pytest-mypy-plugins](https://pypi.org/project/pytest-mypy-plugins/).

File: tests/typecheck/test_basics.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    import factory.django
+
+    class Book(models.Model):
+        ...
+
+    class BookFactory(factory.django.DjangoModelFactory[Book]):
+        ...
+
+    instance = BookFactory.create()
+    reveal_type(instance)  # N: Revealed type is "main.Book"
```
Comment (browniebroke) [inline tests/typecheck/test_basics.yml:13]: Nice one 🙌🏻 

Question: does it also work for `BookFactory()`?

Comment (last-partizan) [reply tests/typecheck/test_basics.yml:13]: No, it doesn't.

In simple cases we just need to add types for `__new__`, but in this case with metaclass i haven't figured out yet how to make it work.

Comment (last-partizan) [reply tests/typecheck/test_basics.yml:13]: @youtux have you managed to get this working in you types?

Comment (youtux) [reply tests/typecheck/test_basics.yml:13]: no, it seems to be a mypy bug: https://github.com/python/mypy/issues/14122

**Comment by Viicos** [comment]
Should `stub`/`stub_batch` be typed to return `StubObject`? Still better than `Any`.

File: tests/typecheck/test_basics.yml (hunk 1 above)
Comment (Viicos) [reply tests/typecheck/test_basics.yml:13]: Here is a workaround until the mypy issue is solved:

```python
T = TypeVar("T")


class BaseFactory(Generic[T]):
    """Factory base support for sequences, attributes and stubs."""

    # Backwards compatibility
    UnknownStrategy = errors.UnknownStrategy
    UnsupportedStrategy = errors.UnsupportedStrategy

    if TYPE_CHECKING:
        def __new__(cls, *args, **kwargs) -> T:
            ...
    else:
        def __new__(cls, *args, **kwargs):
            """Would be called if trying to instantiate the class."""
            raise errors.FactoryError('You cannot instantiate BaseFactory')
```

`mypy` will currently not support this, but will most probably be supported soon (see https://github.com/python/mypy/pull/16020). `pyright` (and thus Pylance) supports it.

Comment (last-partizan) [reply tests/typecheck/test_basics.yml:13]: Updated my code here, it should work now with pyright and when that PR merged - it will work with mypy too.

**Comment by last-partizan** [comment]
> Should `stub`/`stub_batch` be typed to return `StubObject`? Still better than `Any`. Other than that, having this PR merged would be really useful, as you can't really make use of the type stubs as from the implementation side (and not the stubs file), the factory class doesn't inherit from `Generic` or define any `__class_getitem__` class method

I'm not sure. StubObject is generic container, and from type-checking perspective it's empty, so you cannot access any attributes.

**Comment by Viicos** [comment]
> I'm not sure. StubObject is generic container, and from type-checking perspective it's empty, so you cannot access any attributes.

Yes but at least people would know what object it is, instead of `Any` currently (as I believe no type checker can infer the return type considering it is making subcalls to `_generate`)

**Comment by last-partizan** [comment]
@Viicos i think it's better to leave it as is, at least for now. This PR is focused on just basic typing support.

After merging this, you can improve typing for StubFactory.

**Review by Viicos — APPROVED** [review]

**Comment by Viicos** [comment]
Gentle bump @francoisfreitag considering how many people requested it

**Comment by francoisfreitag** [comment]
I don’t see opposition to merging this work. I offered a small fixes as a [PR](https://github.com/last-partizan/factory_boy/pull/3). Once these fixes are correct, and integrated into this PR, I think we can merge it.

In the meantime, @rbarrois’ opinion is very welcome :)

**Comment by last-partizan** [comment]
Oh, I missed that PR. Fixes accepted!

**Review by francoisfreitag — APPROVED** [review]
Looks like a good start. Let’s hope type information PR build up to better document the library :)

I’ll leave the PR sit for a couple more weeks, to give time for others to report their concerns. In the absence of opposition, I’ll merge it early 2024.

File: factory/base.py [hunk 2]
```diff
@@ -506,12 +509,12 @@ def _create(cls, model_class, *args, **kwargs):
...
 
     @classmethod
-    def build(cls, **kwargs):
+    def build(cls, **kwargs) -> T:
         """Build an instance of the associated class, with overridden attrs."""
         return cls._generate(enums.BUILD_STRATEGY, kwargs)
 
     @classmethod
-    def build_batch(cls, size, **kwargs):
+    def build_batch(cls, size, **kwargs) -> List[T]:
```
Comment (Viicos) [inline factory/base.py:517]:
```suggestion
    def build_batch(cls, size: int, **kwargs) -> List[T]:
```

File: factory/base.py [hunk 3]
```diff
@@ -523,12 +526,12 @@ def build_batch(cls, size, **kwargs):
...
 
     @classmethod
-    def create(cls, **kwargs):
+    def create(cls, **kwargs) -> T:
         """Create an instance of the associated class, with overridden attrs."""
         return cls._generate(enums.CREATE_STRATEGY, kwargs)
 
     @classmethod
-    def create_batch(cls, size, **kwargs):
+    def create_batch(cls, size, **kwargs) -> List[T]:
```
Comment (Viicos) [inline factory/base.py:534]:
```suggestion
    def create_batch(cls, size: int, **kwargs) -> List[T]:
```

**Review by Viicos — COMMENTED** [review]
I think it doesn't hurt to type `size` as well

**Comment by francoisfreitag** [comment]
Can you rebase one last time before merging, please? :pray:

**Comment by last-partizan** [comment]
Sure, rebased already, but looks like typing tests aren't running in CI.

Almost ready.

**Comment by last-partizan** [comment]
Got it!

Ready to be merged, but please take another look at github workflow file and tox.ini

File: tox.ini [hunk 4]
```diff
@@ -81,3 +86,7 @@ extras = dev
 
 whitelist_externals = make
 commands = make lint
+
+[testenv:py{38,39,310,311,py39,py310}-typecheck]
```
Comment (francoisfreitag) [inline tox.ini:90]:
```suggestion
[testenv:py{38,39,310,311,312,py39,py310}-typecheck]
```

File: tox.ini [hunk 5]
```diff
@@ -5,6 +5,7 @@ envlist =
     docs
     examples
     linkcheck
+    py{38,39,310,311,py39,py310}-typecheck
```
Comment (francoisfreitag) [inline tox.ini:8]:
```suggestion
    py{38,39,310,311,312,py39,py310}-typecheck
```

**Review by francoisfreitag — APPROVED** [review]

**Comment by rbarrois** [comment]
Thanks for all the hard work! (And sorry for my long absence :/)

I'm going to do a final pass, before merging this; I'm likely going to rebase or squash some of the commits, for easier further maintenance!

**Comment by last-partizan** [comment]
@rbarrois i'm ok with whatever fits you.

If you want to simplify tests or remove py.typed - go ahead.

My main interest - is getting started with typing this package, and then improve types when i encounter problems in my projects.

**Comment by Viicos** [comment]
> Regarding the testing setup, I have met a few issues running it locally; for now, I would rather merge the annotated code without those tests, or with some much simpler ones (e.g using `mypy.reveal_types` directly).

One alternative to `pytest-mypy-plugins` is to do it "manually": you create a python file that you want to test with `mypy` (or any type checker), and use `typing_extensions.assert_type`.

[Here](https://github.com/Viicos/django-autotyping/tree/main/tests/stubbing) is an example of how I do it (test files are under `testfiles`, and `test_stubs.py` is where the actual testing happens).

> Based on PEP 561, I'm not sure we can add this yet: the project is not fully typed.

I'm not sure it should be included either. Without it, IDEs should be smart enough to still use the type definitions, but type checking with the command line will not.

**Comment by rbarrois** [comment]
Thanks again for all this work, I have merged it as #1059 !

**Pull request closed without merging**